### PR TITLE
test/crimson: disable stall-detector on aarch64 to fix asan failures

### PR DESCRIPTION
**Pull request by Svelar**
Ceph crimson uses seastar on debug mode which would enable ASAN, while stall-detector uses glibc backtrace function which would cause ASAN failures on aarch64.
Reason see https://github.com/scylladb/scylladb/issues/15090#issuecomment-1763355796

Because arm ci servers in lab are "elderly", causing stall happened often, this PR is to disable stall-detector until seastar upstream migrated to libunwind, see https://github.com/scylladb/seastar/issues/1878

Seastar does not provide a function to disable stall-detector, at present, increase stall-detector time to avoid stall-detector timeout on arm platform.

fixes: https://tracker.ceph.com/issues/65635

After dozen times of triggering arm make check, all crimson/seastar related tests passed.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by tchaikov** [comment]
the title of this commit reads:

> src/seastar: fix asan error on ARM.

but the body of the commit message is empty. after following the change to seastar, i found https://github.com/ceph/seastar/compare/7d4ae901b5604cb23273a5f0341c8d5f2c553d44...303ec379cb076fa5f36a40d385fb72da6070bd6d#diff-5e5dfd46a396d71e0d68f183d6eb0f3bb5aad37be914323b8e4fea134dafec0c . which still misses the explanation why / how it addresses the ASan error.

**Comment by Svelar** [comment]
> the title of this commit reads:
> 
> > src/seastar: fix asan error on ARM.
> 
> but the body of the commit message is empty. after following the change to seastar submodule, i found [ceph/seastar@7d4ae90...303ec37#diff-5e5dfd46a396d71e0d68f183d6eb0f3bb5aad37be914323b8e4fea134dafec0c](https://github.com/ceph/seastar/compare/7d4ae901b5604cb23273a5f0341c8d5f2c553d44...303ec379cb076fa5f36a40d385fb72da6070bd6d#diff-5e5dfd46a396d71e0d68f183d6eb0f3bb5aad37be914323b8e4fea134dafec0c) . which still misses the explanation why / how it addresses the ASan error.

I know that, still WIP, for running arm make check only.

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test windows

**Comment by tchaikov** [comment]
@Svelar before gets https://github.com/scylladb/seastar/issues/1878 fixed. is it possible to apply a similar change like https://github.com/scylladb/scylladb/pull/15720 to crimson instead of maintaining the patch disabling stall detector in ceph/seastar? IMHO, it's not a joy to maintain another patch in downstream.

**Comment by cyx1231st** [comment]
The root cause looks clear! Should be better to introduce the change inside crimson code instead of the seastar private module, as kefu suggested.

Also include @Matan-B.

**Comment by arm7star** [comment]
For crimson-related test cases, the SEGV problem may be solved by using --blocked-reactor-notify-ms to specify the stall time without modifying the code.

**Comment by tchaikov** [comment]
> For crimson-related test cases, the SEGV problem may be solved by using --blocked-reactor-notify-ms to specify the stall time without modifying the code.

While using `--blocked-reactor-notify-ms` can indeed help mitigate SEGV issues in crimson-related test cases, it's important to note that this approach is a workaround rather than a permanent solution. It's akin to putting a band-aid on a broken leg.

**Comment by Svelar** [comment]
> While using `--blocked-reactor-notify-ms` can indeed help mitigate SEGV issues in crimson-related test cases, it's important to note that this approach is a workaround rather than a permanent solution. It's akin to putting a band-aid on a broken leg.

Yes, you are right. Permanent solution should be migration to libunwind, once upstream fixed, this issue could be "cured".

File: src/test/crimson/seastar_runner.h [hunk 1]
```diff
@@ -71,16 +71,22 @@ struct SeastarRunner {
...
       on_end.reset(new seastar::readable_eventfd);
       return seastar::now().then([this] {
-	begin_signaled = true;
-	[[maybe_unused]] auto r = ::eventfd_write(begin_fd.get(), APP_RUNNING);
-	assert(r == 0);
-	return seastar::now();
+#ifdef __aarch64__
+        seastar::smp::invoke_on_all([] {
+          using namespace std::chrono;
+          eastar::engine().update_blocked_reactor_notify_ms(duration_cast<milliseconds>(10000h));
```
Comment (tchaikov) [inline src/test/crimson/seastar_runner.h:77]: s/eastar/seastar/

does this compile? also this deserves a comment explaining why update the notify timeout and why 10000h.

**Review by tchaikov — COMMENTED** [review]
i am not sure if this compiles.

**Comment by Svelar** [comment]
jenkins test make check

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test api

**Comment by Svelar** [comment]
jenkins test api

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by tchaikov** [comment]
would you please reconsider the suggestion of "also this deserves a comment explaining why update the notify timeout and why 10000h."?

**Comment by Svelar** [comment]
> would you please reconsider the suggestion of "also this deserves a comment explaining why update the notify timeout and why 10000h."?

Updated the commit message, and more likely follow the way that upstream fixed the issue, https://github.com/gusev-p/scylla/commit/f75e949b5fde998f7b331d60a4d6755c4b8470b7

**Comment by Svelar** [comment]
jenkins test api

**Comment by cyx1231st** [comment]
> Updated the commit message, and more likely follow the way that upstream fixed the issue

The commit message looks great, adding a short comment can make the code (expecially as a work-around) much easier to follow without checking the history.

A comment also helps to minimize the risk to drop the work-around unexpectedly.

**Comment by Svelar** [comment]
> The commit message looks great, adding a short comment can make the code (expecially as a work-around) much easier to follow without checking the history.
> 
> A comment also helps to minimize the risk to drop the work-around unexpectedly.

Sounds reasonable, will do ASAP.

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test api

**Comment by Svelar** [comment]
jenkins test make check arm64

**Review by cyx1231st — APPROVED** [review]
LGTM

**Comment by Svelar** [comment]
jenkins test api

**Comment by arm7star** [comment]
jenkins test api

**Comment by Svelar** [comment]
jenkins test make check arm64

File: src/test/crimson/seastar_runner.h [hunk 2]
```diff
@@ -71,16 +71,28 @@ struct SeastarRunner {
...
-	begin_signaled = true;
-	[[maybe_unused]] auto r = ::eventfd_write(begin_fd.get(), APP_RUNNING);
-	assert(r == 0);
-	return seastar::now();
+// FIXME: The stall detector uses glibc backtrace function to
+// collect backtraces, this causes ASAN failures on ARM.
+// For now we just extend timeout duration to 10000h in order to
+// get the same effect as disabling the stall detector which is not provided by seastar.
+// the ticket about migrating to libunwind: https://github.com/scylladb/seastar/issues/1878
+// Will remove once the ticket fixed.
```
Comment (Matan-B) [inline src/test/crimson/seastar_runner.h:79]: Can you please create an issue in our tracker and paste it here as a comment?
`// See: https://tracker.ceph.com/issues/<>`

File: src/test/crimson/seastar_runner.h [hunk 3]
```diff
@@ -71,16 +71,28 @@ struct SeastarRunner {
...
+          seastar::engine().update_blocked_reactor_notify_ms(duration_cast<milliseconds>(10000h));
+        }).get();
+#endif
+        begin_signaled = true;
+        [[maybe_unused]] auto r = ::eventfd_write(begin_fd.get(), APP_RUNNING);
+        assert(r == 0);
+        return seastar::now();
       }).then([this] {
-	return on_end->wait().then([](size_t){});
+        return on_end->wait().then([](size_t){});
```
Comment (Matan-B) [inline src/test/crimson/seastar_runner.h:91]: nit: unrelated whitespace diff

File: src/test/crimson/seastar_runner.h [hunk 4]
```diff
@@ -71,16 +71,28 @@ struct SeastarRunner {
...
+#ifdef __aarch64__
+        seastar::smp::invoke_on_all([] {
+          using namespace std::chrono;
+          seastar::engine().update_blocked_reactor_notify_ms(duration_cast<milliseconds>(10000h));
+        }).get();
+#endif
+        begin_signaled = true;
+        [[maybe_unused]] auto r = ::eventfd_write(begin_fd.get(), APP_RUNNING);
+        assert(r == 0);
+        return seastar::now();
```
Comment (Matan-B) [inline src/test/crimson/seastar_runner.h:89]: nit: ditto, whitespaces - cleaner diff would make it easier to understand/revert in the future

**Review by Matan-B — COMMENTED** [review]
lgtm! few comments

File: src/test/crimson/seastar_runner.h (hunk 2 above)
Comment (Svelar) [reply src/test/crimson/seastar_runner.h:79]: Attached.

File: src/test/crimson/seastar_runner.h (hunk 3 above)
Comment (Svelar) [reply src/test/crimson/seastar_runner.h]: Fixed

File: src/test/crimson/seastar_runner.h (hunk 4 above)
Comment (Svelar) [reply src/test/crimson/seastar_runner.h]: ditto

**Comment by Svelar** [comment]
jenkins test api

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Matan-B** [comment]
I think that both make checks passing is enough to merge this

**Comment by Svelar** [comment]
> I think that both make checks passing is enough to merge this

@Matan-B All passed, could you help merge?

**Comment by Svelar** [comment]
Because it affects ARM CI, IMHO this should be backported, any thought? @cyx1231st @Matan-B
